R/summary_funs.R

```diff
@@ -259,141 +233,178 @@ weighted_summary_means <- function(y_wobs_test, family, wdraws, mu, dis, cl_ref,
+get_stat <- function(summaries, summaries_baseline = NULL,
+                     summaries_fast = NULL, loo_inds = NULL,
+                     y_wobs_test, stat, alpha = 0.1, ...) {
+  mu <- summaries$mu
+  lppd <- summaries$lppd
+  n_full <- length(lppd)
+  n_loo <- if (is.null(loo_inds)) n_full else length(loo_inds)
+  if (all(is.na(lppd)) || all(is.na(y_wobs_test$y_prop %||% y_wobs_test$y))) {
     return(list(value = NA, se = NA, lq = NA, uq = NA))
   }
-
-  if (is.null(wcv)) {
-    ## set default CV fold weights if not given
-    wcv <- rep(1, n)
-  }
-  ## ensure the CV fold weights sum to n_notna
-  wcv <- n_notna * wcv / sum(wcv)
-
   alpha_half <- alpha / 2
   one_minus_alpha_half <- 1 - alpha_half
 
   if (stat %in% c("elpd", "mlpd", "gmpd")) {
-    if (!is.null(lppd.bs)) {
-      value <- sum((lppd - lppd.bs) * wcv, na.rm = TRUE)
-      value.se <- weighted.sd(lppd - lppd.bs, wcv, na.rm = TRUE) *
-        sqrt(n_notna)
+    if (is.null(summaries_baseline)) {
+      lppd_baseline <- 0
     } else {
-      value <- sum(lppd * wcv, na.rm = TRUE)
-      value.se <- weighted.sd(lppd, wcv, na.rm = TRUE) *
-        sqrt(n_notna)
+      lppd_baseline <- summaries_baseline$lppd
+    }
+    if (n_loo < n_full) {
+      # subsampling difference estimator (Magnusson et al., 2020)
+      srs_diffe <- .srs_diff_est_w(y_approx = summaries_fast$lppd - lppd_baseline,
+                                   y = (lppd - lppd_baseline)[loo_inds],
+                                   y_idx = loo_inds)
+      value <- srs_diffe$y_hat
+      # combine estimates of var(y_hat) and var(y)
+      value_se <- sqrt(srs_diffe$v_y_hat + srs_diffe$hat_v_y)
+    } else {
+      # full LOO estimator
+      value <- sum(lppd - lppd_baseline, na.rm = TRUE)
+      value_se <-sd(lppd - lppd_baseline, na.rm = TRUE) * sqrt(n_full)
     }
     if (stat %in% c("mlpd", "gmpd")) {
-      value <- value / n_notna
-      value.se <- value.se / n_notna
+      value <- value / n_full
+      value_se <- value_se / n_full
       if (stat == "gmpd") {
         value_gmpd <- exp(value)
-        # Delta method:
-        value_gmpd.se <- value.se * value_gmpd
+        # delta method
+        value_gmpd_se <- value_se * value_gmpd
       }
     }
-  } else if (stat %in% c("mse", "rmse")) {
+  } else if (stat %in% c("mse", "rmse", "R2")) {
     y <- y_wobs_test$y_prop %||% y_wobs_test$y
-    if (!all(y_wobs_test$wobs == 1)) {
-      wcv <- wcv * y_wobs_test$wobs
-      wcv <- n_notna * wcv / sum(wcv)
+    wobs <- y_wobs_test$wobs
+    wobs <- n_full * wobs / sum(wobs)
+    if (!is.null(summaries_baseline)) {
+      mu_baseline <- summaries_baseline$mu
     }
-    if (stat == "mse") {
-      if (!is.null(mu.bs)) {
-        value <- mean(wcv * ((mu - y)^2 - (mu.bs - y)^2), na.rm = TRUE)
-        value.se <- weighted.sd((mu - y)^2 - (mu.bs - y)^2, wcv,
-                                na.rm = TRUE) /
-          sqrt(n_notna)
+    # Use normal approximation for mse and delta method for rmse and R2
```

Comment:
Is this supposed to refer to the standard error estimation method or the CI method? The first part ("normal approximation") refers to a CI method, but the second part ("delta method") to a standard error estimation method.